controllers/src/main/java/com/epam/esm/controllers/GiftCertificateController.java

```
@@ -0,0 +1,132 @@
+     * @param giftCertificate is a GiftCertificateTagsWrapper entity included new giftCertificate with associated tags;
+     */
+    @PostMapping
+    public void createNewGiftCertificate(@RequestBody GiftCertificateTagsWrapper giftCertificate) {
+        giftCertificateService.saveNewGiftCertificate(giftCertificate.getGiftCertificate(), giftCertificate.getTags());
+    }
+
+    /**
+     * Updates the giftCertificate with associated tags, which will be created if
+     * they do not exist and are associated with the updated giftCertificate;
+     *
+     * @param giftCertificate is an entity to be updated;
+     * @throws ResourceNotFoundException if the resource being updated does not found;
+     */
+    @PutMapping
+    public void updateGiftCertificate(@RequestBody GiftCertificateTagsWrapper giftCertificate) throws ResourceNotFoundException {
+        Long giftCertificateId = giftCertificate.getGiftCertificate().getId();
+        Optional<GiftCertificateDto> giftCertificateDto = giftCertificateService.findById(giftCertificateId);
+        if (giftCertificateDto.isPresent()) {
+            giftCertificateService.updateGiftCertificate(giftCertificate.getGiftCertificate(), giftCertificate.getTags());
+        } else {
+            String resource = " (Gift certificate id = " + giftCertificateId + ")";
+            throw new ResourceNotFoundException(resource);
+        }
+    }
+
+    /**
+     * Finds giftCertificate by id and creates a corresponding dto object;
+     *
+     * @param id is a unique field of giftCertificate;
+     * @return GiftCertificateDto is the generated corresponding dto object based on the found giftCertificate
+     * @throws ResourceNotFoundException if the requested giftCertificate is not found;
+     */
+    @GetMapping("{id}")
+    public GiftCertificateDto getGiftCertificate(@PathVariable Long id) throws ResourceNotFoundException {
+        Optional<GiftCertificateDto> giftCertificateDto = giftCertificateService.findById(id);
```

I suppose you don't store certificate with id=null, that way better to throw an exception that id must be passed or just return null without service bean.